### ISSUE TITLE
LED colors aren't set in d-bus

### ISSUE THREAD

**Issue by AlexanderAmelkin** [openbmc/openbmc#3430]
Since commit openbmc/phosphor-dbus-interfaces@a1e70a7af by Vishwanatha Subbanna (see https://gerrit.openbmc-project.xyz/#/c/openbmc/phosphor-dbus-interfaces/+/1136/ ), there is a property 'Color' for leds.

However, it looks like nothing is setting that property for any of the supported platforms. So any led always has it in d-bus set to Unknown.

It is unclear to us what was the intention for adding that property. It has not been explained in the commit log for the commit. It is also unclear what could be the correct way to set that property. phosphor-led-manager operates on groups. A single led can be a part of multiple groups, so 'led.yaml' in phosphor-led-manager looks like a wrong place to attach a color property to leds. Each leds used in led.yaml is called by the name taken from Linux devicetree. However, the bindings for leds in Linux devicetree do not imply color information as well.

Are there any hints on how to set that property the right way? Can you imagine any real use cases for the color property? Maybe we should just ditch it completely? It's been 2 years since it has been introduced and nobody seems to have cared...

**Comment by vishwabmc** [comment]
@AlexanderAmelkin When it was first written, `palette` was introduced for future extensions and there was no use-case to utilize that back then. 

Are we having a use-case now to utilize that property and asking what is the right way to do so ?

I believe we could hook something into: 
 https://github.com/openbmc/phosphor-led-manager/blob/master/manager.cpp#L163

I understand we currently don't have the color information for the group. But we may need to implement that into the group definition so we could use that to drive the Physical LED.

I will be happy to be part of this discussion.

**Comment by AlexanderAmelkin** [comment]
Yes, you're right. We're asking "what is the right way to do so". We just don't like a property with "Unknown" value hanging around. We've asked back on Nov 6 in the mailing list for the proper method to initialize it from the knowledge that we have about the hardware, and nobody answered (you were on a CC). Hence, we've decided to add our own initialization. It's great that it have drawn your attention. :-)

I wouldn't call it a true 'runtime' configuration, but technically it, of course, is runtime. If you have a compile-time method in mind to set LED colors and not have it showing "Unknown", please share it.

As for the use case, I could think of a UI rendering the appropriate color based on this property.

**Comment by AlexanderAmelkin** [comment]
> So by default, when the LED is off, it would not have any color

This is completely wrong, IMO. There is a `State` property for that. An `Off` green led doesn't loose its color just due to being off.

The patchset at https://gerrit.openbmc-project.xyz/16080 allows for setting initial colors (as in "hardware properties") for single-color LEDs yet preserve the mechanism that you describe for RGB LEDs (although it won't be functional due to reasons described below). You just don't set any color for RGB LEDs in the configuration file and they will remain "Unknown" until set by any other software. Please note though that there is no software to set that property. **phosphor-led-manager** operates in terms of LED groups, it uses led nodes in sysfs to turn leds on and off, and has no idea of "color". 

Please also note that Linux sysfs doesn't have the idea of color either. Even RGB leds would be represented as three separate physical LEDs, each having a distinct single color property: red, green or blue. An application would then use the sysfs `brightness` attribute to set color proportions. I believe actually that exactly this was the reason for only having base LED colors in the Palette interface.

**Comment by vishwabmc** [comment]
I will ask @jk-ozlabs to see how a multi-color led ( if at all one is needed for server platform ) will be presented via sysfs. Is it one entry with the intelligence built into the respective driver -or- will that be presented as 3 different leds and thus expecting the applications to drive what is needed to achieve the color combination. if latter, then I see there is really no usecase for that color palette.

**Comment by AlexanderAmelkin** [comment]
@vishwabmc, currently the led subsystem in kernel doesn't have support for color mixing. There are only interfaces for brightness and blinking control.

**Comment by vishwabmc** [comment]
@jk-ozlabs and I talked about this and he pointed me to : https://github.com/torvalds/linux/blob/master/Documentation/leds/leds-class.txt#L46

Here is the gist of what Jeremy and I talked:

> Current LED driver will represent RGB as 3 different LEDs
> Jeremy liked the idea of presenting a unified interface for RGB than 3 different but not something that can be worked on now
> Also talked about having an abstraction layer at application 

without an abstraction later, the best way is to follow : "devicename:colour:function" in the device tree.
Existing DT needs to be updated to rename the led to have the color.

As an example: today's "fan0' will need to be renamed as "fan0:blue" ( or whatever color ) in DT
Since the physical led daemon is getting spun in response to uevent, the color will anyway be provided as uevent and hence this eliminates any runtime env file. Needs a minor change in leds.

@AlexanderAmelkin please let me know if you have any feedback.

**Comment by AlexanderAmelkin** [comment]
@vishwabmc, we'd be fine with this approach if it worked. Alas, Linux cuts off anything before the colon so if you have `led_name:color` in DT you end up with a `/sys/class/leds/blue`.

In fact we were thinking about using DT, especially in view of the already present color information in the names of some leds (e.g. `id_blue`, `alarm_red`), but thought that this could be unreliable.

On the other hand, we could set up a specification for led naming in DT for openbmc and demand them end in color matching one of those from `Palette` like `id_blue` or `fan0_blue`.

Also, what do you mean by `function`? Can you give an example?

**Comment by vishwabmc** [comment]
@AlexanderAmelkin I run ubuntu and have this 

```vishwa@vishwa:/sys/class/leds$ ls
phy0-led  tpacpi::power  tpacpi::standby  tpacpi::thinklight  tpacpi::thinkvantage
vishwa@vishwa:/sys/class/leds$ 

So the function is `power / standby / thinklight / thinkvantage`.  It does not have the color info here and hence you see 2 semicolons in succession. But if it had color, then it would be a:b:c

This maps to : https://github.com/torvalds/linux/blob/master/Documentation/leds/leds-class.txt#L46

Since I see it working on ubuntu, this should still work for us in BMC

```

**Comment by AlexanderAmelkin** [comment]
@vishwabmc, thank you, we've missed the part that it needs to be a `label` property, not the DT node name itself. It indeed works. It appears though, that there are going to be changes to this part in the Linux upstream:

https://www.spinics.net/lists/linux-leds/msg10525.html

The changes, though, are going to keep compatibility with the current behavior.
I think it is indeed the good way to go. We will change our patch accordingly.

**Comment by AlexanderAmelkin** [comment]
Posting a link to the proposed change based on this discussion:
https://gerrit.openbmc-project.xyz/c/openbmc/phosphor-led-sysfs/+/16080

**Comment by vishwabmc** [comment]
This has been already merged..

**Comment by AlexanderAmelkin** [comment]
Indeed. It's been merged as openbmc/phosphor-led-sysfs@97ddb72359663b5ca9d8e0c22588f0022f729ced
